### PR TITLE
Feature: Adding Token Processing Message for Old token transfers

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: api/scripts/add_message.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import os
```
Comment (krishan711) [inline api/scripts/add_message.py:1]: much better file name needed please

File: api/scripts/add_message.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+@click.option('-s', '--start-block-number', 'startBlockNumber', required=True, type=int)
+@click.option('-e', '--end-block-number', 'endBlockNumber', required=True, type=int)
+@click.option('-b', '--batch-size', 'batchSize', required=False, type=int, default=100)
+async def add_message(startBlockNumber: int, endBlockNumber: int, batchSize: int):
+    database = Database(f'postgresql://{os.environ["DB_USERNAME"]}:{os.environ["DB_PASSWORD"]}@{os.environ["DB_HOST"]}:{os.environ["DB_PORT"]}/{os.environ["DB_NAME"]}')
+    sqsClient = boto3.client(service_name='sqs', region_name='eu-west-1', aws_access_key_id=os.environ['AWS_KEY'], aws_secret_access_key=os.environ['AWS_SECRET'])
+    workQueue = SqsMessageQueue(sqsClient=sqsClient, queueUrl='https://sqs.eu-west-1.amazonaws.com/097520841056/notd-work-queue')
+    
+    await database.connect()
+    cache =set()
```
Comment (krishan711) [inline api/scripts/add_message.py:31]:
```suggestion
    cache = set()
```

File: api/scripts/add_message.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        start = currentBlockNumber
+        end = min(currentBlockNumber + batchSize, endBlockNumber)
+        logging.info(f'Working on {start} to {end}...')
+        async with database.transaction():
+            await database.connect()
+            query = TokenTransfersTable.select()
+            query = query.where(TokenTransfersTable.c.blockNumber >= start)
+            query = query.where(TokenTransfersTable.c.blockNumber < end)
+            query = query.where(TokenTransfersTable.c.registryAddress.in_(
+                TokenTransfersTable.select()
```
Comment (krishan711) [inline api/scripts/add_message.py:43]: this looks like the oppsite of what we want -are you only getting tokens that already exist? We want tokens that dont exist currentkly. Also this looks very expensive, does it run fast locally?

File: api/scripts/add_message.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                .with_only_columns([TokenTransfersTable.c.registryAddress])
+                    .group_by(TokenTransfersTable.c.registryAddress)
+            ))
+            tokenTransfersToChange = [token_transfer_from_row(row) async for row in database.iterate(query=query)]
+            for tokenTransfer in tokenTransfersToChange:
+                cache.add((tokenTransfer.registryAddress,tokenTransfer.tokenId))
```
Comment (krishan711) [inline api/scripts/add_message.py:49]: this is not a great way to use a cache. what you should do here is check if the registry+tokenId is in the cache and then 1) if it is: skip 2) if its not: send the message + add to cache

File: api/scripts/add_message.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        logging.info(f'Found {len(cache)} unique tokens between {startBlockNumber} and {end}')
+        currentBlockNumber = currentBlockNumber + batchSize
+    collectionSet = set()
+    for i in range(len(cache)):
+        registryAddress = list(list(cache)[i])[0]
+        tokenId = list(list(cache)[i])[1]
+        await workQueue.send_message(message=UpdateTokenMetadataMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message())
+        collectionSet.add(registryAddress)
+    for address in collectionSet:
+        await workQueue.send_message(message=UpdateCollectionMessageContent(address=address).to_message())
```
Comment (krishan711) [inline api/scripts/add_message.py:59]: you should have two separate caches here i think once you have done the suggestion above

File: api/scripts/add_message.py (hunk 1 above)
Comment (Femi-Ogunkola) [reply api/scripts/add_message.py:1]: add_queue_message?

File: api/scripts/add_message.py (hunk 3 above)
Comment (Femi-Ogunkola) [reply api/scripts/add_message.py:43]: So the approach I took here was to get all unique tokens between two block numbers and then add them all to a set

File: api/scripts/add_queue_message.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import os
```
Comment (krishan711) [inline api/scripts/add_queue_message.py:1]: you need to make this file name something more relevant. like `schedule_processing_for_tokens_from_old_transfers.py`

Comment (krishan711) [reply api/scripts/add_queue_message.py:1]: pretty much all scripts add messages to queues which is why this is not a great name

File: api/scripts/add_queue_message.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+@click.command()
+@click.option('-s', '--start-block-number', 'startBlockNumber', required=True, type=int)
+@click.option('-e', '--end-block-number', 'endBlockNumber', required=True, type=int)
+@click.option('-b', '--batch-size', 'batchSize', required=False, type=int, default=100)
+async def add_message(startBlockNumber: int, endBlockNumber: int, batchSize: int):
+    database = Database(f'postgresql://{os.environ["DB_USERNAME"]}:{os.environ["DB_PASSWORD"]}@{os.environ["DB_HOST"]}:{os.environ["DB_PORT"]}/{os.environ["DB_NAME"]}')
+    #sqsClient = boto3.client(service_name='sqs', region_name='eu-west-1', aws_access_key_id=os.environ['AWS_KEY'], aws_secret_access_key=os.environ['AWS_SECRET'])
+    #workQueue = SqsMessageQueue(sqsClient=sqsClient, queueUrl='https://sqs.eu-west-1.amazonaws.com/097520841056/notd-work-queue')
+    sqsClient = boto3.client(service_name='sqs', region_name='us-east-1', aws_access_key_id=os.environ['AWS_KEY'], aws_secret_access_key=os.environ['AWS_SECRET'])
+    workQueue = SqsMessageQueue(sqsClient=sqsClient, queueUrl='https://sqs.us-east-1.amazonaws.com/113848722427/test1')
```
Comment (krishan711) [inline api/scripts/add_queue_message.py:30]: this looks wrong - we can make the queue name and region name environment variables if this makes things easier to test

File: api/scripts/add_queue_message.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    while currentBlockNumber < endBlockNumber:
+        start = currentBlockNumber
+        end = min(currentBlockNumber + batchSize, endBlockNumber)
+        logging.info(f'Working on {start} to {end}...')
+        async with database.transaction():
+            await database.connect()
+            query = TokenTransfersTable.select()
+            query = query.where(TokenTransfersTable.c.blockNumber >= start)
+            query = query.where(TokenTransfersTable.c.blockNumber < end)
+            query = query.where(TokenTransfersTable.c.registryAddress.in_(
```
Comment (krishan711) [inline api/scripts/add_queue_message.py:44]: what happened here.. why is this in_?

File: api/scripts/add_queue_message.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                TokenTransfersTable.select()
+                .with_only_columns([TokenTransfersTable.c.registryAddress])
+                    .group_by(TokenTransfersTable.c.registryAddress)
+            ))
+            tokenTransfersToChange = [token_transfer_from_row(row) async for row in database.iterate(query=query)]
+            for tokenTransfer in tokenTransfersToChange:
+                if (tokenTransfer.registryAddress, tokenTransfer.tokenId) in cache:
+                    pass
```
Comment (krishan711) [inline api/scripts/add_queue_message.py:52]: `continue` and then you can remove the else below

File: api/scripts/add_queue_message.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                else:
+                    cache.add((tokenTransfer.registryAddress,tokenTransfer.tokenId))
+                    await workQueue.send_message(message=UpdateTokenMetadataMessageContent(registryAddress=tokenTransfer.registryAddress, tokenId=tokenTransfer.tokenId).to_message())
+                    if tokenTransfer.registryAddress in registryCache:
+                        pass
```
Comment (krishan711) [inline api/scripts/add_queue_message.py:57]: same here, and nicer to tab this out

File: api/scripts/add_queue_message.py (hunk 8 above)
Comment (Femi-Ogunkola) [reply api/scripts/add_queue_message.py:44]: So the in_ was to get a list of all the registry token pairs that had appeared once more

Comment (krishan711) [reply api/scripts/add_queue_message.py:44]: once more than what?

Comment (krishan711) [reply api/scripts/add_queue_message.py:44]: if we removed this what would happen?

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    currentBlockNumber = startBlockNumber
+    while currentBlockNumber < endBlockNumber:
+        start = currentBlockNumber
+        end = min(currentBlockNumber + batchSize, endBlockNumber)
+        logging.info(f'Working on {start} to {end}...')
+        async with database.transaction():
+            await database.connect()
+            query = TokenTransfersTable.select()
+            query = query.where(TokenTransfersTable.c.blockNumber >= start)
+            query = query.where(TokenTransfersTable.c.blockNumber < end)
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:41]: try to exclude any tokens that are already in the token_metadatas table

Comment (krishan711) [reply api/scripts/schedule_processing_for_tokens_from_old_transfers.py:45]: pls do update this when you get a sec 👍

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    await database.connect()
+    cache = set()
+    registryCache = set()
+    currentBlockNumber = startBlockNumber
+    while currentBlockNumber < endBlockNumber:
+        start = currentBlockNumber
+        end = min(currentBlockNumber + batchSize, endBlockNumber)
+        logging.info(f'Working on {start} to {end}...')
+        async with database.transaction():
+            await database.connect()
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:42]: this looks wrong

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            query = TokenTransfersTable.select()
+            query = query.where(TokenTransfersTable.c.blockNumber >= start)
+            query = query.where(TokenTransfersTable.c.blockNumber < end)
+            query = query.where(sqlalchemy.tuple_(TokenTransfersTable.c.registryAddress,TokenTransfersTable.c.tokenId).not_in(
+                    TokenMetadataTable.select()
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:47]: the tabbing here is wierd.

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+                        .with_only_columns([TokenMetadataTable.c.registryAddress,TokenMetadataTable.c.tokenId])
+                            .group_by(TokenMetadataTable.c.registryAddress,TokenMetadataTable.c.tokenId)))
+            tokenTransfersToChange = [token_transfer_from_row(row) async for row in database.iterate(query=query)]
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:50]: could you print `query` here and paste what you see into a comment on this PR pls

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            for tokenTransfer in tokenTransfersToChange:
+                if (tokenTransfer.registryAddress, tokenTransfer.tokenId) in cache:
+                    continue
+                cache.add((tokenTransfer.registryAddress,tokenTransfer.tokenId))
+                await workQueue.send_message(message=UpdateTokenMetadataMessageContent(registryAddress=tokenTransfer.registryAddress, tokenId=tokenTransfer.tokenId).to_message())
+                if tokenTransfer.registryAddress in registryCache:
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:56]: where are you adding things into the cache?

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+            query = query.where(TokenTransfersTable.c.blockNumber < end)
+            query = query.where(sqlalchemy.tuple_(TokenTransfersTable.c.registryAddress,TokenTransfersTable.c.tokenId).not_in(
+                    TokenMetadataTable.select()
+                        .with_only_columns([TokenMetadataTable.c.registryAddress,TokenMetadataTable.c.tokenId])
+                            .group_by(TokenMetadataTable.c.registryAddress,TokenMetadataTable.c.tokenId)))
+            tokenTransfersToChange = [token_transfer_from_row(row) async for row in database.iterate(query=query)]
+            for tokenTransfer in tokenTransfersToChange:
+                if (tokenTransfer.registryAddress, tokenTransfer.tokenId) in cache:
+                    continue
+                cache.add((tokenTransfer.registryAddress,tokenTransfer.tokenId))
```
Comment (krishan711) [inline api/scripts/schedule_processing_for_tokens_from_old_transfers.py:54]: spaces between items in a tuple please (its above too)

File: api/scripts/schedule_processing_for_tokens_from_old_transfers.py (hunk 14 above)
Comment (Femi-Ogunkola) [reply api/scripts/schedule_processing_for_tokens_from_old_transfers.py:50]: SELECT tbl_token_transfers.id, tbl_token_transfers.transaction_hash, tbl_token_transfers.registry_address, tbl_token_transfers.from_address, tbl_token_transfers.to_address, tbl_token_transfers.token_id, tbl_token_transfers.value, tbl_token_transfers.gas_limit, tbl_token_transfers.gas_price, tbl_token_transfers.gas_used, tbl_token_transfers.block_number, tbl_token_transfers.block_hash, tbl_token_transfers.block_date 
FROM tbl_token_transfers 
WHERE tbl_token_transfers.block_number >= :blockNumber_1 AND tbl_token_transfers.block_number < :blockNumber_2 AND ((tbl_token_transfers.registry_address, tbl_token_transfers.token_id) NOT IN (SELECT tbl_token_metadatas.registry_address, tbl_token_metadatas.token_id 
FROM tbl_token_metadatas GROUP BY tbl_token_metadatas.registry_address, tbl_token_metadatas.token_id))

**Pull request merged**